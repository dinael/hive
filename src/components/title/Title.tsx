import { FC } from 'react'
import styles from './Title.module.scss'

export type TitleProps = {
  level: 0 | 1 | 2 | 3 | 4 | 5 | 6
  size?: 'xs' |'s' | 'm' | 'l' | 'xl' | 'xxl'
  text: string
  children?: any
  ellipsis?: boolean
  a11yLength?: boolean
  className?: string
}

export const Title: FC<TitleProps> = ({
  level,
  size = 'm',
  text,
  children,
  ellipsis = false,
  a11yLength = false,
  className,
  ...props
}: TitleProps ) => {

  level = level < 0 || level > 6 ? 2 : level

  const Tag = level === 0 ? 'p' : 'h' + level as keyof JSX.IntrinsicElements

  const titleStyles = `
    ${styles['title-component']}
    ${styles[size]}
    ${className || ''}
    ${ellipsis ? styles['ellipsis'] : ''}
    ${a11yLength ? styles['a11y-length'] : ''}
  `

  return (
    <Tag
      className={titleStyles}
      {...props}>
        {text}
      {children}
    </Tag>
  )
}

export default Title
