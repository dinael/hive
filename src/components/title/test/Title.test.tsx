import React from 'react'
import { describe, test, expect } from 'vitest'
import { render, screen } from '@testing-library/react'
import { Title, TitleProps } from '../Title'

describe('Title', () => {
  const defaultProps: TitleProps = {
    level: 3,
    size: 'xxl',
    text: 'Example title',
    ellipsis: true,
    a11yLength: true,
    className: 'title-component',
    children: null,
  }

  const notAssignableLevel = 7;

  test('renders snapshot', () => {
    const { container } = render(<Title {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  test('renders with the correct tag name', () => {
    const { container } = render(<Title {...defaultProps} />)
    const titleElement = container.querySelector('h3')
    expect(titleElement?.tagName).toBe('H3')
    expect(container).toMatchSnapshot()
  });

  test('renders with the correct tag name when level is 0', () => {
    const { container } = render(<Title {...defaultProps} level={0} />)
    const titleElement = container.querySelector('p')
    expect(titleElement?.tagName).toBe('P')
    expect(container).toMatchSnapshot()
  });

  test('renders with the correct tag name when level is out range', () => {
    const { container } = render(<Title {...defaultProps} level={notAssignableLevel} />)
    const titleElement = container.querySelector('h2')
    expect(titleElement?.tagName).toBe('H2')
    expect(container).toMatchSnapshot()
  });

  test('renders with the correct text', () => {
    const { container } = render(<Title {...defaultProps} />)
    const titleElement = container.querySelector('h3')
    expect(titleElement?.textContent).toBe('Example title')
    expect(screen.getByText('Example title'))
  });

  test('renders with the correct extra class name', () => {
    const { container } = render(<Title {...defaultProps} />)
    const titleElement = container.querySelector('h3')
    expect(titleElement?.className).toContain('title-component')
  });

  test('renders with the correct size', () => {
    const { container } = render(<Title {...defaultProps} />)
    const titleElement = container.querySelector('h3')
    expect(titleElement?.className).toContain('xxl')
  });

  test('renders with the correct ellipsis', () => {
    const { container } = render(<Title {...defaultProps} />)
    const titleElement = container.querySelector('h3')
    expect(titleElement?.className).toContain('ellipsis')
  });

  test('renders with the correct text length to a11y', () => {
    const { container } = render(<Title {...defaultProps} />)
    const titleElement = container.querySelector('h3')
    expect(titleElement?.className).toContain('a11y-length')
  });

})


