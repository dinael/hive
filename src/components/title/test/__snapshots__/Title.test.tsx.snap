// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Title > renders snapshot 1`] = `
<div>
  <h3
    class="
        title-component
        xxl
        title-component
        ellipsis
        a11y-length
      "
  >
    Example title
  </h3>
</div>
`;

exports[`Title > renders with the correct tag name 1`] = `
<div>
  <h3
    class="
        title-component
        xxl
        title-component
        ellipsis
        a11y-length
      "
  >
    Example title
  </h3>
</div>
`;

exports[`Title > renders with the correct tag name when level is 0 1`] = `
<div>
  <p
    class="
        title-component
        xxl
        title-component
        ellipsis
        a11y-length
      "
  >
    Example title
  </p>
</div>
`;

exports[`Title > renders with the correct tag name when level is out range 1`] = `
<div>
  <h2
    class="
        title-component
        xxl
        title-component
        ellipsis
        a11y-length
      "
  >
    Example title
  </h2>
</div>
`;
